src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```
@@ -0,0 +1,2164 @@
+                        }
+
+                        _sessionHandle.DangerousAddRef(ref ignore);
+                    }
+                }
+            }
+        }
+
+        private bool GetServerCredentialsFromCache(
+            Uri uri,
+            out uint serverAuthScheme,
+            out NetworkCredential serverCredentials)
+        {
+            serverAuthScheme = 0;
+            serverCredentials = null;
+
+            NetworkCredential cred = null;
+
+            lock (_credentialCacheLock)
+            {
+                foreach (uint authScheme in s_authSchemePriorityOrder)
+                {
+                    cred = _credentialCache.GetCredential(uri, s_authSchemeStringMapping[authScheme]);
+                    if (cred != null)
+                    {
+                        serverAuthScheme = authScheme;
+                        serverCredentials = cred;
+
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private void SaveServerCredentialsToCache(Uri uri, uint authScheme, ICredentials serverCredentials)
+        {
+            string authType = s_authSchemeStringMapping[authScheme];
+            Debug.Assert(!string.IsNullOrEmpty(authType));
+
+            NetworkCredential cred = serverCredentials.GetCredential(uri, authType);
+            if (cred != null)
+            {
+                lock (_credentialCacheLock)
+                {
+                    try
+                    {
+                        _credentialCache.Add(uri, authType, cred);
+                    }
+                    catch (ArgumentException)
+                    {
+                        // The credential was already added.
+                    }
+                }
+            }
+        }
+
+        private async void StartRequest(object obj)
+        {
+            RequestState state = obj as RequestState;
```

Could be ```(RequestState)obj``` instead of ```obj as RequestState```.  Since we require obj to be a RequestState at this point, that would help with diagnostics if this ever failed, as it'd resultin an InvalidCastException from here rather than a NullReferenceException from later in the method if obj was some other non-null type.